libraries/AP_AIS/AP_AIS.cpp

```diff
@@ -0,0 +1,765 @@
+
+    // log the raw infomation
+    if ((_options & AIS_OPTIONS_LOG_DECODED) != 0) {
+        const struct log_AIS_msg1 pkt{
+            LOG_PACKET_HEADER_INIT(LOG_AIS_MSG1),
+            time_us      : AP_HAL::micros64(),
+            type         : type,
+            repeat       : repeat,
+            mmsi         : mmsi,
+            nav          : nav,
+            rot          : rot,
+            sog          : sog,
+            pos_acc      : pos_acc,
+            lon          : lon,
+            lat          : lat,
+            cog          : cog,
+            head         : head,
+            sec_utc      : sec_utc,
+            maneuver     : maneuver,
+            raim         : raim,
+            radio        : radio
+        };
+        AP::logger().WriteBlock(&pkt, sizeof(pkt));
+    }
+
+    uint16_t index;
+    if (!get_vessel_index(mmsi, index, lat, lon)) {
+        // no room in the vessel list
+        return true;
+    }
+
+    // mask of flags that we receive in this message
+    const uint16_t mask = ~(AIS_FLAGS_POSITION_ACCURACY | AIS_FLAGS_VALID_COG | AIS_FLAGS_VALID_VELOCITY | AIS_FLAGS_VALID_TURN_RATE | AIS_FLAGS_TURN_RATE_SIGN_ONLY);
+    uint16_t flags = _list[index].info.flags & mask; // clear all flags that will be updated
+    if (pos_acc) {
+        flags |= AIS_FLAGS_POSITION_ACCURACY;
+    }
+    if (cog < 36000) {
+        flags |= AIS_FLAGS_VALID_COG;
+    }
+    if (sog < 1023) {
+        flags |= AIS_FLAGS_VALID_VELOCITY;
+    }
+    if (sog == 1022) {
+        flags |= AIS_FLAGS_HIGH_VELOCITY;
+    }
+    if (rot > -128) {
+        flags |= AIS_FLAGS_VALID_TURN_RATE;
+    }
+    if (rot == 127 || rot == -127) {
+        flags |= AIS_FLAGS_TURN_RATE_SIGN_ONLY;
+    } else {
+        rot = powf((rot / 4.733f),2.0f) / 6.0f;
+    }
+
+    _list[index].info.lat = lat; // int32_t [degE7] Latitude
+    _list[index].info.lon = lon; // int32_t [degE7] Longitude
+    _list[index].info.COG = cog; // uint16_t [cdeg] Course over ground
+    _list[index].info.heading = head; // uint16_t [cdeg] True heading
+    _list[index].info.velocity = sog; // uint16_t [cm/s] Speed over ground
+    _list[index].info.flags = flags; // uint16_t Bitmask to indicate various statuses including valid data fields
+    _list[index].info.turn_rate = rot; // int8_t [cdeg/s] Turn rate
+    _list[index].info.navigational_status = nav; // uint8_t Navigational status
+    _list[index].last_update_ms = AP_HAL::millis();
+
+    return true;
+}
+
+bool AP_AIS::decode_static_and_voyage_data(const char *payload)
+{
+    if (strlen(payload) != 71) {
+        return false;
+    }
+
+    char call_sign[8];
+    char name[21];
+    char dest[21];
+
+    uint8_t repeat      = get_bits(payload, 6, 7);
+    uint32_t mmsi       = get_bits(payload, 8, 37);
+    uint8_t ver         = get_bits(payload, 38, 39);
+    uint32_t imo        = get_bits(payload, 40, 69);
+               get_char(payload, call_sign, 70, 111);
```

Comment:
odd indenting here..

Reply:
I found this nicer when checking I had all the bit ranges correct (the last two args). That is to say it is odd indenting is by design but I can change it if your really don't like it,

Reply:
It's ok, it's your baby.